Spike/idg modifications

idg10 opened the pull request:
First two chapters currently have more or less the structure and examples I want.

HowardvanRooijen committed:
Convert to GitHub flavoured Markdown

HowardvanRooijen committed:
Merge pull request #1 from endjin/feature/polyglot-notebooks

Convert to GitHub flavoured Markdown

HowardvanRooijen committed:
Update project location from codeplex to GitHub

HowardvanRooijen committed:
Update links

HowardvanRooijen committed:
reinstatedl

idg10 committed:
Add experimental notebook

idg10 committed:
Decent first draft of ch01

idg10 committed:
WIP on ch02

idg10 committed:
More progress on Ch02

idg10 committed:
Ch01 and ch02 structure now in place

Moved some stuff into placeholders in later chapters.

HowardvanRooijen commented on content/01_WhyRx.md:
If you add the context of broadband networking equipment, then it nicely ties into the next set of examples.

HowardvanRooijen commented on content/01_WhyRx.md:
You could use an AIS example - i.e. two vessels exhibiting suspicious behaviour, as a way to tie back to the previous list too.

HowardvanRooijen commented on content/01_WhyRx.md:
I'm wondering if we could work a cloud standard like https://cloudevents.io/ into either this section, or a demo in the book...

HowardvanRooijen replied:
Also wondering whether we could mention MQTT for IOT scenarios (not just for completeness, but for SEO juice for the introtorx.com site)

HowardvanRooijen commented on content/01_WhyRx.md:
I'm also wondering if you need to mention `System.Reactive` in this sentence too.. I notice it's referenced later on, but what it is and how it related hasn't really been explained.

HowardvanRooijen commented on content/01_WhyRx.md:
We haven't really explained the link between Reactive Extensions / Rx.NET / Rx and System.Reactive.

HowardvanRooijen commented on content/01_WhyRx.md:
Where does Channels fit into this ecosystem?

HowardvanRooijen commented on content/01_WhyRx.md:
What about backpressure? It's been added to RxJava, but is missing from Rx.NET. That's the only other big flashing "things I'd expect to just work"

idg10 replied:
Addressed with update to ch2

HowardvanRooijen commented on content/02_KeyTypes.md:
```diff
@@ -1,488 +1,531 @@
+
+Suppose you want to write a program that watches some folder on a filesystem, and performs automatic processing any time something in that folder changes. For example, web developers often to trigger automatic rebuilds of their client side code when they save changes in the editor so they can quickly see the effect of their changes. Often in these cases, filesystem changes often come in bursts—te ...
+
+So what we really want to report is those moments when everything goes quiet after a flurry of activity. The following code defines an Rx operator that detects and reports such things. If you're new to Rx (which seems likely if you're reading this) it probably won't be instantly obvious how this works, so I'll walk through it.
```
is it worth specifying it's a custom Rx operator - otherwise people might assume it's an native one. Actually at this point have you actually described what an operator is?

idg10 committed:
Lifetime Management chapter content moved into other chapters

idg10 committed:
Work in progress on Creating Observable Sequences

idg10 committed:
First full draft of Creating Observable Sequences

idg10 committed:
Add hot/cold section to Key Types

idg10 committed:
Add PART 2 break

idg10 committed:
First full draft of Filtering

idg10 committed:
Remove Inspection chapter. Start on Transformation

idg10 committed:
Transformation WIP

idg10 committed:
Feature complete draft of Transformation chapter

idg10 committed:
Aggregation WIP

idg10 committed:
Feature-complete draft of Aggregation chapter

idg10 committed:
Structurally complete Partitioning chapter

idg10 committed:
Reworked Concat in Combining Sequences

HowardvanRooijen commented on content/09_CombiningSequences.md:
````diff
@@ -2,68 +2,103 @@
+```
 
-Usage of `Concat` is familiar. Just like `Catch` or `OnErrorResumeNext`, we pass the continuation sequence to the extension method.
+Since of `Concat` is an extension method, we can invoke it as a method on any sequence, passing the second sequence in as the only argument:
````
Typo "of"

idg10 committed:
Basic structure of Combining Sequences

idg10 committed:
Re-order part 3 content

idg10 committed:
Partial update of scheduling chapter

idg10 committed:
More scheduler updates

idg10 committed:
Feature-complete Leaving Observable chapter

idg10 committed:
Feature complete Timing chapter

idg10 committed:
Feature complete Error Handling chapter

HowardvanRooijen committed:
Fix missing ` which were breaking pandoc

HowardvanRooijen committed:
Fix formatting / typos

idg10 committed:
Remove old Hot and Cold chapter

* Add the one bit of interesting content it had into Ch02
* Add a new Publishing Operators chapter, to discuss the few operators that didn't have a good home yet, and which were previously in the Hot and Cold chapter
* Added forward refs to this new chapter in the places I was using Publish and RefCount, and also a ref to PublishLast from the LastAsync section in Ch05

idg10 committed:
Fix double closing backtick

idg10 committed:
Change to code fence for stack trace

Turns out the book rendering doesn't like deeply nested bullet lists

idg10 committed:
Add line breaks to AIS output to placate pandoc

idg10 committed:
Minor clarification in disposal description

idg10 committed:
Add Observable.Defer section

idg10 committed:
Experiment with <figure>

Seeing if this lets us specify both a caption and a description.

idg10 committed:
Move Dump example one chapter earlier

idg10 committed:
Minor tweaks to Ch15 publishing operators

idg10 committed:
Testing chapter WIP

idg10 committed:
Ch14 more error handling updates

idg10 committed:
Draft of testing chapter

idg10 committed:
Renumber testing chaper

idg10 committed:
Move remaining content out of combiningsequences and into other chapters

idg10 committed:
Update appendix A, put all remaining Disposable content in appendix B

idg10 committed:
Remove more spurious tab characters

idg10 committed:
Remove Dispelling Myths chapter

This all seemed only tangentially related to Rx, and also felt rather
outdated.

idg10 committed:
Fix appendix a title

idg10 committed:
First full Appendix D draft

idg10 committed:
Fix missing file extension on Appendix D

idg10 committed:
Replace unicode with maths expressions

idg10 committed:
Remove spurious unicode thin spaces

...accidentally copied from another page. Apparently our book build process can't cope with the Unicode thin space character.

idg10 committed:
Fix another silly LaTeX mistake

idg10 committed:
Fix appendix B title

idg10 committed:
Remove spurious backticks from Ch16

idg10 committed:
Update Appendix C

idg10 committed:
Initial containerised book PDF build

idg10 committed:
Ch00 update after read through

idg10 committed:
Ch01 update after read through

idg10 committed:
Ch02 updates after read through

idg10 committed:
Ch03 partial read through

idg10 committed:
Ch03 partial read through

idg10 committed:
Ch03 completed read-through

idg10 committed:
Update Part2 after read through

Also moved some notes to Part3, to deal with when I reach that part.

idg10 committed:
Remove spuriously added build outputs

Update build scripts

idg10 committed:
Ch05 update after read through

idg10 committed:
Ch06 update after partial read through

idg10 committed:
Ch06 updates after read through

idg10 committed:
Handle various TODOs

idg10 committed:
Deal with some TODOs, hide others.

Left some in as comments so we can get a servicable book produced

idg10 committed:
Fix chapter ref

idg10 committed:
Update build script after chapter reorder

Also remove spurious copy step - the pandoc commands all now put the output where it's wanted so there's no need for the copy.

idg10 committed:
Fix appendix D title

idg10 committed:
Ch07 partial read through

idg10 committed:
Ch07 partial updated for read through

idg10 committed:
Ch07 updated after read through

idg10 committed:
Update Ch08 after read through

idg10 committed:
Ch09 partial read through

idg10 committed:
Ch09 updates after read through

idg10 committed:
Ch10 update after read through

idg10 committed:
Ch11 partial read through

idg10 committed:
Ch11 update after read through

idg10 committed:
Ch12 update after read through

idg10 committed:
Ch13 update after read through

idg10 committed:
Ch14 completed read through

idg10 committed:
Ch15 update after read through

idg10 committed:
Ch16 updates after read through

idg10 committed:
Appendix A update after read through

idg10 committed:
Appendix B update after read through

idg10 committed:
Appendix C update after read through

idg10 committed:
Remove bits of foreword that were messing up the formatting

idg10 committed:
Add missing space in heading

idg10 committed:
Appendix D partial read

idg10 committed:
Appendix D updates after read through

HowardvanRooijen commented on content/00_Foreword.md:
```diff
@@ -1,185 +1,105 @@
+generously allowed the Rx.NET project to make use of his content, enabling this
+new edition to come into existence.
+
+I would also like to recognize the people that made this book
```
I would restructure this. Put your acknowledgements first and then include "acknowledgements from the 1st edition". I think you've earned the right!

idg10 replied:
I've flipped the order  in the 2nd paragraph. I'm still leading with the fact that Lee wrote the first edition though. Do you think I should bump that lower down too? That feels like it might be a step to far to me.

HowardvanRooijen commented on content/02_KeyTypes.md:
```diff
@@ -1,488 +1,627 @@
+
+This is Rx's _compositional_ approach, and it is how we normally use Rx. We use a mixture of operators, combined (_composed_) in a way that produces the effect we want.
+
+But how does this particular combination produce the effect we want? There are a few ways we could get the behaviour that we're looking for from a `Quiescent` operator, but the basic idea of this particular implementation is that it keeps count of how many events have happened recently, and then produces a result every time that number drops back to zero. The `oustanding` variable refers to the ` ...
```
Typo "oustanding"

HowardvanRooijen commented on content/02_KeyTypes.md:
```diff
@@ -1,488 +1,627 @@
+
+That middle section is the messiest, but it's also most representative of the kind of activity this operator is designed to deal with. Remember, the whole point here is that we're expecting to see flurries of activity, and if those represents filesystem activity, they will tend to be slightly chaotic in nature, because storage devices don't always have entirely predictable performance characteris ...
+
+With this measure of recent activity in hand, we can spot the end of bursts of activity by watching for when `oustanding` drops back to zero, which is what the observable referred to by `zeroCrossing` in the code above does. (That's just using the `Where` operator to filter out everything except the events where `outstanding`'s current value returns to zero.)
```
Typo "oustanding"

HowardvanRooijen commented on content/02_KeyTypes.md:
```diff
@@ -1,488 +1,627 @@
+
+This calls `obs.OnNext` 10,000 times, but it executes these calls as individual tasks to be run on the thread pool. The thread pool is designed to be able to execute work in parallel, and that's a a problem here because nothing here ensures that one call to `OnNext` completes before the next begins. We've broken the rule that says we must wait for each call to `OnNext` to return before calling ei ...
+
+This rule in which we must wait for `OnNext` to return is tricky and subtle. It's perhaps less obvious than the others, because there's no equivalent rule for `IEnumerable<T>`—the opportunity to break this rule only arises when the source pushes data into the application. You might look at the example above and think "well who would do that?" However, multithreading is just an easy way to show th ...
```
I thought "re-entrancy" was "reentrancy"

idg10 replied:
Opinion is divided on the matter. Oxford University Press's "Oxford Dictionary of English" hyphenates it, as does the Cambridge University Press's "Cambridge Dictionary.' Merriam Webster does not, nor does the Chicago Style Manual, or the Associated Press.

But it's not merely a US/UK thing because O'Reilly's house style is also re-entrant/re-entrancy, so that's how it appears in my books, and that's the habit I'm in.

HowardvanRooijen commented on content/06_Transformation.md:
```diff
@@ -0,0 +1,351 @@
+
+The Rx version of the example we're currently examining is in fact one of these unusual cases where each of the sequences emits items as soon as it can. Logically speaking, all of the nested `IObservable<char>` sequences are in progress concurrently. The result is a mess because each of the observable sources here attempts to produce every element as quickly as the source can consume them. The fa ...
+
+![An Rx marble diagram showing 7 observables. The first illustrates the Range operator producing the values 1 through 5. These are colour coded as follows: green, blue, yellow, red, and pink. These colours correspond to observables further down in the diagram, as will be described shortly. The items in this first observable are not evenly spaced—the 2nd value immediately follows the 1st, but ther ...
```
Typo "lamba"

HowardvanRooijen commented on content/06_Transformation.md:
```diff
@@ -0,0 +1,351 @@
+
+![An Rx marble diagram showing 7 observables. The first illustrates the Range operator producing the values 1 through 5. These are colour coded as follows: green, blue, yellow, red, and pink. These colours correspond to observables further down in the diagram, as will be described shortly. The items in this first observable are not evenly spaced—the 2nd value immediately follows the 1st, but ther ...
+
+We can make a small tweak to prevent the child sequences all from trying to run at the same time. (This also uses `Observable.Repeat` instead of the rather indirect route of constructing a `string` and then calling `ToObserable` on that. I did that in earlier examples to emphasize the similarity with the LINQ to Objects example, but you wouldn't really do it that way in Rx.)
```
Typo "ToObserable"

HowardvanRooijen commented on content/16_TestingRx.md:
```diff
@@ -1,846 +1,722 @@
+
+* That the `Price` property has prices added to it as the model produces them
+* That the sequence is subscribed to on the ThreadPool
+* That the `Price` property is updated on the Dispatcher i.e. the sequenceis observed on the Dispatcher
```
Typo "sequenceis"

HowardvanRooijen commented on content/16_TestingRx.md:
```diff
@@ -1,846 +1,722 @@
+
+### Start(Func<IObservable<T>>)
+
+There are three overloads to `Start`, which are used to start an observablesequence at a given time, record the notifications it makes and dispose of the subscription at a given time. This can be confusing at first, as the parameterless overload of `Start` is quite unrelated. These three overloads return an `ITestableObserver<T>` which allows you to record the notifications from an observable seq ...
```
Is this a typo, or should it have formatting? "observablesequence"

HowardvanRooijen commented on content/D_AlgebraicUnderpinnings.md:
```diff
@@ -0,0 +1,600 @@
+
+Where it gets slightly more weird (and where Rx departs from the original definition of `amb`) is when more than one of an ambiguous function's constituents produces a result. In McCarthy's theoretical formulation, the ambiguous function effectively produces all of the results as possible outputs. (This is technically known as _nondeterministic_ computation, although that name can be misleading:  ...
+
+[Rx's `Amb`](09_CombiningSequences.md#amb) provides the behaviour described in the cases where either none of the inputs produces anything, or exactly one of them does. Howeever, it makes no attempt to support non-deterministic computation, so its handling of the case where multiple constituents are able to produce value is oversimplified, but then McCarthy's `amb` was first and foremost an analy ...
```
Typo "Howeever"

HowardvanRooijen commented on content/C_UsageGuidelines.md:
```diff
@@ -0,0 +1,26 @@
+- Avoid creating side effects. If you really can't avoid it, don't bury the side effects in callbacks for operators designed to be use functionally such as `Select` or `Where`. be explicit by using the `Do` operator.
+- Where possible, prefer `Observable.Create` to subjects as a means of defining new Rx sources.
+- Avoid creating your own implementations of the `IObservable<T>` interface. Use `Observable.Create` (or subjects if you really need to).
+- Avoid creating your own implementations of the `IObserver<T>` interface. Favor using the `Subscribe` extension method overloads instead.
```
Not sure if you're using en-gb or en-us spellings, but "Favor"

idg10 replied:
For the record, en-gb. Because the majority of what Lee wrote did too.

IIRC, this appendix was almost unaltered, so I guess this was always inconsistent.

HowardvanRooijen commented on content/00_Foreword.md:
```diff
@@ -1,185 +1,105 @@
+
+I would also like to recognize the people that made this book
+possible. Crucial to the first edition of the book, in addition to the author, [Lee Campbell](https://leecampbell.com/), were: 
+[James Miles](http://enumeratethis.com/), 
```
It might be worth checking these sites still exist - I know all the blogs.msdn.com have gone - maybe we should point to LinkedIn profiles instead, especially as many people have left Microsoft since the 1st edition.

I also wonder if we should point to the existance of A Little History of Reaqtor is people want a deeper insights into Rx and the teams and research projects involved.

HowardvanRooijen replied:
http://enumeratethis.com/ no longer exists

HowardvanRooijen commented on content/01_WhyRx.md:
```diff
@@ -1,241 +1,121 @@
+
+## Why Rx?
+
+Users want timely information. If you're waiting for a parcel to arrive, live reports of the delivery van's progress give you more freedom than a suspect 2 hour delivery window. Financial applications depend on continuous streams of up-to-date data. We expect our phones and computers to provide us with all sorts of important notifications. And some applications simply can't work without live info ...
```
"market data"?

idg10 replied:
Is that not already subsumed by "financial applications"?

HowardvanRooijen commented on content/01_WhyRx.md:
Are both definitions essentially implementing the Future or Promise pattern? Both are defining a query, but neither are executed until you either ask for a value (IEnumerable) or Subscribe to observe (IObservable)?

HowardvanRooijen commented on content/01_WhyRx.md:
Is there anything here to address around performance? Folks love LINQ, but it's got a bad reputation around performance. How does this play out for Rx?

Also should there be a link here to the Algebraic underpinnings appendix?

HowardvanRooijen replied:
Do we need to make the distinction that IObservable lives in the BCL and thus is a .NET Primative, and then later where we talk about `System.Reactive` we can explain the relationship to `IObservable`?

HowardvanRooijen commented on content/01_WhyRx.md:
I'd quite like to use the word "telemetry" in here somewhere as a use case.

HowardvanRooijen commented on content/01_WhyRx.md:
"when our processing" - is this the effect of applying business rules and domain transformations to turn it from data to information or "actionable insights"? I think this is quite a key concept that people are missing and why Rx is so important in real time scenarios. People understand how we go from data to "actionable insights" with python / spark / ELT / ETL but it's really not a concept we talk about WRT Rx.

HowardvanRooijen replied:
Also feels like "signal to noise" ratio plays a part in this...

HowardvanRooijen commented on content/01_WhyRx.md:
Would this be where something like System.Reaqtive might enable checkpointing support, because it adds the notion of operator state into the paradigm?

HowardvanRooijen commented on content/01_WhyRx.md:
Is there something around the general trend that people think reactive programming is complex - do Rx queries suffer from the same problem as LINQ and RegExps in that if you're coming to a complex statement cold, it can be hard to deciper?

HowardvanRooijen commented on content/02_KeyTypes.md:
```diff
@@ -1,488 +1,627 @@
+
+# Key types
+
+Rx is a powerful framework that can greatly simplify code that responds to events. But to write good Reactive code you have to understand the basic concepts. The fundamental building block of Rx is an interface called `IObservable<T>`. Understanding this, and its counterpart `IObserver<T>`, is the key to success with Rx.
```
Do you need to mention this is a BCL type?

idg10 replied:
Once we get out of this intro material, and into the section dedicated to the specific interfaces, I say this:

> that it is part of the .NET runtime libraries

So I already do. (Except I always avoid the term BCL because it's ambiguous. For example, according to the ECMA spec, `IObservable<T>` very definitely isn't in the BCL, because the BCL is actually a very narrow subset of the types in the runtime class libraries.)

mwadams commented on content/00_Foreword.md:
```diff
@@ -1,185 +1,105 @@
+---
+
+Reactive programming is not a new concept. Any kind of user interface development
+necessary involves code that responds to events. Languages like Smalltalk, Delphi
```
necessary => necessarily

mwadams commented on content/00_Foreword.md:
```diff
@@ -1,185 +1,105 @@
+concurrency.
+
+The .NET runtime libraries have included the `IObservable<T>` and
+`IObserver<T>` interfaces that representing the core concept of reactive programming
```
representing => represent

mwadams commented on content/00_Foreword.md:
```diff
@@ -1,185 +1,105 @@
+The .NET runtime libraries have included the `IObservable<T>` and
+`IObserver<T>` interfaces that representing the core concept of reactive programming
+for well over a decade now. The Reactive Extensions for .NET are effectively a library of implementations of these
+interfaces. Rx.NET implementation first appeared
```
Maybe either `The RX.NET implementation first appeared` or `RX.NET first appeared`

mwadams commented on content/00_Foreword.md:
```diff
@@ -1,185 +1,105 @@
+sentence.)
+
+If you have never used Rx before, it _will_ change the way you design and build software.
+It provides a well thought out abstraction for a fundamentally important idea in computing—sequences
```
Maybe a `:` instead of an `em-dash`

mwadams commented on content/01_WhyRx.md:
```diff
@@ -1,241 +1,121 @@
+
+In short, our systems need to react when interesting things happen.
+
+Live information streams are a basic, ubiquitous element of computer systems. Despite this, they are often a second class citizen in programming languages. Most languages support sequences of data through something like an array, which presumes that the data is sitting in memory ready for our code to read at its leisure. If your application deals with events, arrays might work for historical data ...
```
I'm not sure this is the right place for it, but it might be worth addressing the fact that Streams were the historical "one thing after another" abstraction, and why that has difficulties.

mwadams commented on content/01_WhyRx.md:
I would replace the `-` with a full stop and reduce run-on sentence size.

mwadams commented on content/01_WhyRx.md:
Again, I'd remove the `-`` and add a full stop to reduce run-on sentence length.

mwadams commented on content/01_WhyRx.md:
Maybe:

For example, some message queuing technologies such as MSMQ are by definition sequential, and thus might look like a good fit for Rx. However, they are often chosen for their transaction handling support. Rx does not have any direct way to surface transaction semantics, so in scenarios that require this you might be better off just working directly with the relevant technology's API.

mwadams commented on content/01_WhyRx.md:
I realise this is all style stuff, but I'd definitely go for more full stops and fewer `-` masquerading as a full stop.

Mainly because I write a lot of `-` masquerading as a full stop myself, and it always seems better (and more direct/emphatic) with a real full stop when I go back and revise it.

mwadams commented on content/00_Foreword.md:
```diff
@@ -1,185 +1,105 @@
+Reactive programming is not a new concept. Any kind of user interface development
+necessary involves code that responds to events. Languages like Smalltalk, Delphi
+and the .NET languages have popularized reactive or event-driven programming paradigms.
+Architectural patterns such as CEP (Complex Event Processing), and
```
Do we need links out to some suitable definition for CEP, CQRS or even Smalltalk and Delphi in the previous paragrpah? Antiques like us might know them, but does everyone?

mwadams commented on content/00_Foreword.md:
```diff
@@ -1,185 +1,105 @@
+
+The event driven paradigm allows for code to be invoked without the need for breaking
+encapsulation or applying expensive polling techniques. There are many common ways to implement this, including
+the Observer pattern, events exposed directly in the language (e.g. C#) or
```
Would this benefit from a link, too? (Including the C# events - it could link to the MS documentation)

mwadams reviewed (commented):
Review comments for 00 and 01

HowardvanRooijen commented on content/02_KeyTypes.md:
```diff
@@ -1,488 +1,627 @@
+
+This kind of _cold_-then-_hot_ source can present a problem if we want to attach multiple subscribers. If the source starts providing events as soon as subscription occurs, then that's fine for the very first subscriber: it will receive any events that were backed up waiting for us to start. But if we wanted to attach multiple subscribers, we've got a problem: that first subscriber might receive  ...
+
+In these cases, we really want some way to rig up all our subscribers before kicking things off. We want subscription to be separate from the act of starting. By default, subscribing to a source implies that we want it to start, but rx defines a specialised interface that can give us more control: [`IConnectableObservable<T>`](https://github.com/dotnet/reactive/blob/f4f727cf413c5ea7a704cdd4cd9b4a ...
```
Typo lowercase "rx"

HowardvanRooijen commented on content/02_KeyTypes.md:
```diff
@@ -1,488 +1,627 @@
+* `MoveNext` could throw an exception
+* `MoveNext` could return `false` to indicate that you've reached the end of the collection
+
+These three outcomes correspond precisely to the three methods defined by `IObserver<T>`. We could describe these in slightly more abstract terms:
```
Would you say it's a more "pure" / less overloaded description of the MoveNext behaviour - making each method more Single Responsiblity?

idg10 replied:
I think there's an interesting point in there trying to get out, but don't think those are quite the right words.

Having watched most of the old Meijer, Beckman, and Bart videos from the dawn of Rx, this certainly doesn't seem to be how they thought about it. Their perspective, as I understand it, is that this split into three pieces is very much not at attempt to make Rx more pure; on the contrary, it's an attempt to make Rx exactly model what `IEnumerable<T>` is. It's a simple consequence of the duality: these three facets that were combined into `MoveNext`+`Current` become explicitly visible in `IObserver<T>`.

Eric and Bart have independently talked about what purified versions of `IEnumerable<T>` and `IObservable<T>` would look like. (They've both done videos on this, and Bart did a blog about it too.) They both show a simpler model where you don't do exceptions, and where you model the end of a sequence just as a distinct value that comes out of the sequence. (So this makes both models more pure. You can represent the whole thing as just functions at that point—you don't need interfaces.)

And I think "single responsibility" is not the right abstraction here for two reasons. First, the three `IObserver<T>` methods are tightly coupled: their joint usage is governed by a grammar, so none of these methods is used in isolation. Second, these are literally an alternative expression of what `IEnumerable<T>` and `IEnumerator<T>` signify, and the fact that these three aspects are not fully separated there means that they aren't truly separable in `IObserver<T>`. There may be three distinct methods but they are connected.

So if I were to attempt to express what I think you're getting at, I'd say it's that being push-based makes the possibility of exceptions an explicit part of the model. (The explicit separation between "next value" and "end" was in fact already present in `IEnumerable<T>` with the split between `Current` and `MoveNext`). It forces us to see something that was already there but which was much easier to overlook in the `IEnumerable<T>` model. But equally, there are facets of the abstraction which are inherent in the `IEnumerable<T>` construction that are not directly visible in the Rx formulation, which is why we need to describe the grammar in order to have a complete picture of what `IObserver<T>` is. There are some aspects of that grammar that are simply inherent to the type in the `IEnumerable<T>` formulation.

So I think the moral of the story here is that the .NET type system is not powerful enough to fully express everything about the relationship between the provider and consumer of a sequence. Both `IEnumerable/IEnumerator<T>` and `IObservable/IObserver<T>` capture only their own subsets of the whole story. Different bits of that provider/consumer relationship are explicit and implicit in these two formulations. Neither is really more or less pure or overloaded than the other. Each collapses different parts of the whole.

But I'm not sure the introduction is the place for this sort of type system philosophy. I would expect it to confuse more than it would help someone trying to get their head around this for the first time. Possibly it could go in the appendix on the mathematical underpinnings, although this is more "musings" than "foundations".

HowardvanRooijen commented on content/02_KeyTypes.md:
````diff
@@ -1,488 +1,627 @@
+    `GoUntilStopped.Stop`
+```
+ 
+ Our `GoUntilStopped.Stop` method needs to wait for `OnNext` to return before calling `OnCompleted`. But notice that the `OnNext` method can't return until our `Stop` method returns. We've managed to create a deadlock with single-threaded code!
````
Should this line be indented?

idg10 replied:
Yes. If you look at the two likes above it, the 2nd and 3rd both indent slightly more. I did this to convey the nesting.

HowardvanRooijen commented on content/02_KeyTypes.md:
```diff
@@ -1,488 +1,627 @@
+
+In this case it's not all that hard to fix: we could modify `Stop` so it just sets the `running` field to `false`, and then move the call to `OnComplete` into the `Go` method, after the `for` loop. But more generally this can be a hard problem to fix, and it's one of the reasons for using the `System.Reactive` library instead of just attempting to implement `IObservable<T>` and `IObserver<T>` dir ...
+
+So as long as you're using Rx to build the observable sources you need, you get to depend on these rules in your callbacks that receive the events, and it's mostly Rx's problem to keep to the rules. So the main effect of these rules is that it makes life simpler for code that consumes events.
```
It feels like you're repeating the previous paragraph here.

HowardvanRooijen commented on content/02_KeyTypes.md:
```diff
@@ -1,488 +1,627 @@
+
+This formally captures the basic idea: there can be any number of calls to `OnNext` (maybe even zero calls), that occur in sequence, followed by either an `OnError` or an `OnComplete`, but not both, and there must be nothing after either of these.
+
+One last point: sequences may be infinite. This is true for `IEnumerable<T>`—it's perfectly possible for an enumerator to return `true` every time `MoveNext` is returned, in which case a `foreach` loop iterating over it will never reach the end. It might choose to stop (with a `break` or `return`), or some exception that did not originate from the enumerator might cause the loop to terminate, but ...
```
The best example I remember of trying to define this behaviour, was, I believe Don Box who talked about streaming live video - it never ends.

idg10 replied:
I wonder how that plays to a modern crowd. I used to use conventional television (the only form of streaming live video for which Don's statement is true) as a practical example of a "hot" source. But one of the things I learned when helping a former endjin employee get up to speed with Rx and Reaqtor is that this is apparently not how the younger generation thinks about video.

Video is a thing you click, you watch it, you can pause it, and it stops at the end. While they're technically aware of this old-fangled rather inflexible type of TV where you don't get to decide what you can watch or when, they see that as an aberration, not the default.

HowardvanRooijen commented on content/02_KeyTypes.md:
````diff
@@ -1,488 +1,627 @@
+(OnNext)*(OnError|OnComplete)?
+```
+
+More subtly, observable sources are allowed to do nothing at all. In fact there's a built-in implementation to save developers from the effort of writing a source that does nothing: if you call `Observable.Never<int>()` it will return an `IObservable<int>`, and if you subscribe to that, it will never call any methods on your observer. This might not look immediately useful—it is logically equival ...
````
Can you come up with a concrete, real world, example?

idg10 replied:
I have tried. It's surprisingly hard, because in practice these things are typically buried somewhere at the deepest level of something like the `Quiescent` example, at which point they're very hard to relate to anything real. So what I've come up with is a concrete but made up example masquerading as a real world one.

HowardvanRooijen commented on content/02_KeyTypes.md:
```diff
@@ -1,488 +1,627 @@
+
+You already know from the rules of `IObserver<T>` that a call to either `OnComplete` or `OnError` denotes the end of a sequence. We passed an `IObserver<T>` to `IObservable<T>.Subscribe`, and now the subscription is over. But what if we want to stop the subscription earlier?
+
+I mentioned earlier that the `Subscribe` method returns an `IDisposable`, which enables us to cancel our subscription. Perhaps we only subscribed to a source because our application opened some window showing the status of some process, and we wanted to update the window to reflect that's process's progress. If the user closes that window, we no longer have any use for the notifications. And alth ...
```
battery life, or power consumption?

HowardvanRooijen commented on content/02_KeyTypes.md:
```diff
@@ -1,488 +1,627 @@
+
+### Cancelling Subscriptions may be Slow or Even Ineffectual
+
+`Dispose` won't necessarily take effect instantly. Obviously it will take some non-zero amount of time in between your code calling into `Dispose`, and the `Dispose` implementation reaching the point where it actually does something. Less obviously, some observable sources may need to do non-trivial work to shut things down.
```
So is the behaviour much more like CancellationToken? But it pre-dates CancellationToken semantics being present in the BCL?

idg10 replied:
> So is the behaviour much more like CancellationToken

More like that than what? What's the alternative that you have in mind?

As far as I'm aware, cancellation of any asynchronous activity is normally like this. Back when I taught for DevelopMentor, I used to talk about the fact that cancellation of an asynchronous operation is in itself inherently an asynchronous operation. If the thing you want to cancel is truly asynchronous that means it's operating on its own schedule (e.g., because the work is actually being done by a separate piece of hardware in the computer) so any attempt by the CPU to stop it will generally take some time to effect.

I left DevelopMentor in about 2005, so that was definitely pre-Rx!

But it was already old news then. Kernel mode device drivers in Windows NT (which I'd worked on since 1993) have a cancellation model that works in much the same way. And I would guess that was probably inspired by the much older operating systems Dave Cutler worked on before joining Microsoft. (I suspect you'd find IBM probably had a much earlier commercial implementation of this concept, most likely in OS/360.)

So this is just what cancellation is like. CancellationToken, Rx unsubscription, cancellable IRPs in the NT kernel...they all look like this.

HowardvanRooijen commented on content/02_KeyTypes.md:
```diff
@@ -1,488 +1,627 @@
+
+`Dispose` won't necessarily take effect instantly. Obviously it will take some non-zero amount of time in between your code calling into `Dispose`, and the `Dispose` implementation reaching the point where it actually does something. Less obviously, some observable sources may need to do non-trivial work to shut things down.
+
+A source might create a thread to be able to monitor for and report whatever events it represents. (That would happen with the filesystem source shown above when running on Linux on .NET 7, because the `FileSystemWatcher` class itself creates its own thread on Linux.) It might take a while for the thread to detect that it is supposed to shut down.
```
.NET 8.0 ?

HowardvanRooijen commented on content/02_KeyTypes.md:
```diff
@@ -1,488 +1,627 @@
+
+This sometimes catches people out. If you need to be able to cancel some process that you are observing but you need to be able to observe everything it does up until the point that it stops, then you can't use unsubscription as the shutdown mechanism. As soon as you've called `Dispose`, the `IObservable<T>` that returned that `IDisposable` is no longer under any obligation to tell you anything.  ...
+
+One subtle consequence of this is that if an observable source reports an error after a subscriber has unsubscribed, that error might be lost. A source might call `OnError` on its observer, but if that's a wrapper provided by Rx relating to a subscription that has already been disposed, it just ignores the exception. So it's best to think of early unsubscription as inherently messy, a bit like ab ...
```
Am I correct in remembering that this situation caused problems for the ReactiveUI team? Is it worth re-using any of your writings on that issue to weave it into this section?

idg10 replied:
What I wrote here was directly informed by the work I did to understand the issue the ReactiveUI team had. This _is_ the output of that work.

HowardvanRooijen commented on content/02_KeyTypes.md:
````diff
@@ -1,488 +1,627 @@
+    () => Console.WriteLine("OnCompleted"));
+```
+
+We're calling `Subscribe` on the observable returned by `Where`. When we do that, it will in turn call `Subscribe` on the  the `IObservable<int>` returned by `GetSource` (stored in the `source` variable). So there is in effect a chain of subscriptions here. (We only have access to the `IDisposable` returned by `filtered.Subscribe` but the object that returns will be storing the `IDisposable` that ...
````
Duplicate " the  the "

HowardvanRooijen commented:
00 & 01 reviewed

HowardvanRooijen reviewed (changes requested):
00 & 01 reviewed

HowardvanRooijen commented on content/03_CreatingObservableSequences.md:
```diff
@@ -0,0 +1,1274 @@
+        // finished handling one event before it delivers the next. The Mac,
+        // Windows, and Linux implementations are all significantly different,
+        // so it would be unwise to rely on anything not guaranteed by the
+        // documentation. (As it happens, the Win32 implementation on .NET 7
```
True for .NET 8.0?

HowardvanRooijen commented on content/03_CreatingObservableSequences.md:
```diff
@@ -0,0 +1,1274 @@
+
+### Observable.Return
+
+One of the simplest factory methods is `Observable.Return<T>(T value)`, which you've already seen in the `Quiescent` example in the preceding chapter. This method takes a value of type `T` and returns an `IObservable<T>` which will produce this single value and then complete. In a sense, this _wraps_ a value in an `IObservable<T>`; it's conceptually similar to writing `new T[] { value }`, in that ...
```
Is another example like `Task.FromResult()`?

HowardvanRooijen commented on content/03_CreatingObservableSequences.md:
```diff
@@ -0,0 +1,1274 @@
+
+### Observable.Empty
+
+Sometimes it can be useful to have an empty sequence. .NET's `Enumerable.Empty<T>()` does this for `IEnumerable<T>`, and Rx has a direct equivalent in the form of `Observable.Empty<T>()`, which returns an empty `IObservable<T>`. We need to provide the type argument because there's no value from which the compiler can infer the type.
```
Is this similar to `Task.CompletedTask` in Async?

HowardvanRooijen commented on content/03_CreatingObservableSequences.md:
````diff
@@ -0,0 +1,1274 @@
+IObservable<string> never = Observable.Never<string>();
+```
+
+It might not seem obvious why this could be useful. It tends to be used in places where we use observables to represent time-based information. Sometimes we don't actually care what emerges from an observable; we might care only _when_ something (anything) happens. For example, in the preceding chapter, the `Quiescent` example used the `Buffer` operator, which works over two observable sequences: ...
````
I still feel this needs a concrete, real world example.

HowardvanRooijen commented on content/03_CreatingObservableSequences.md:
Are there any concerns or points of note about usage `ConfigureAwait()` in this scenario?

HowardvanRooijen commented on content/03_CreatingObservableSequences.md:
What about .NET Interactive / Polyglot Notebooks?

HowardvanRooijen commented on content/03_CreatingObservableSequences.md:
add cs to the code fence

HowardvanRooijen commented on content/03_CreatingObservableSequences.md:
Type or Types?
In other places you've used `IObservable<T>` in the heading rather than encoding the <> - should the approach be consistent?

idg10 replied:
Turns out our current toolchain just destroys anything inside angle bracket in a heading. So I guess I'd better use entities everywhere!

idg10 replied:
Ah - in fact it destroys them unless they are in `code` backticks. So in fact what I should probably do is be consistent about whether to use the code style in headings.

HowardvanRooijen commented on content/03_CreatingObservableSequences.md:
I'm wondering whether we should swap out `@"c:\temp"` with `System.IO.Path.GetTempPath`

idg10 replied:
Definitely not, but I'm going to change the path because I evidently made the code suggest something I very much didn't intend it to.

I'm changing it to `c:\incoming` to get away from any suggestion of "temp folder". The reason I wrote `c:\temp` is just because it's where I tend to do experimental stuff—if I want to play around with `FileSystemWatcher` to understand how it works, I'd use the `c:\temp` folder. I do that specifically because that's **my** folder, and not a system folder. (That's why `System.IO.Path.GetTempPath` would definitely be the wrong thing. The problem with using the "real" temp folder is that all sorts of things use that, making it a problematic choice for experimentation.)

So in my head that signified "sandbox".

But I've switched to `c:\incoming` because that's not going to exist on most machines, but is suggestive of some sort of "drop" location that you might want to observe.

HowardvanRooijen commented on content/03_CreatingObservableSequences.md:
Should that be "Ahead of Time"?

HowardvanRooijen commented on content/03_CreatingObservableSequences.md:
Cloud or Serverless environments?

HowardvanRooijen commented on content/03_CreatingObservableSequences.md:
There seems to be two styles of lists in the document - `*` and `-` not sure why we haven't standardised

idg10 replied:
You'd need to ask Lee.

I didn't think that replacing half of the existing examples would be a useful application of time.

HowardvanRooijen commented on content/03_CreatingObservableSequences.md:
Missing . at the end of the sentence.

HowardvanRooijen commented on content/03_CreatingObservableSequences.md:
It feels like it needs sample output for completeness...

idg10 replied:
Since it produces output identical to the previous code I've just stated that this is the case, instead of actually showing the same output a 2nd time

HowardvanRooijen commented on content/03_CreatingObservableSequences.md:
A real-world scenario would be useful

idg10 replied:
Since you've selected text spanning both the `Subject` and `ReplaySubject` sections, it's not clear which bit you're talking about. This is mostly from `Subject`, but since that does in fact have a real example (extracting messages from a queue and delivering them into the world of Rx), I guess you're referring to `ReplaySubject<T>`.

I have never found a use for `ReplaySubject<T>`. There have been times when I thought I had, but later discovered it wasn't a good idea. I don't really know what it's for. (In fact I have a slight suspicion it was ill-conceived.) I'm talking about it for completeness.

HowardvanRooijen commented on content/03_CreatingObservableSequences.md:
A real-world scenario would be useful

HowardvanRooijen commented on content/03_CreatingObservableSequences.md:
A real-world scenario would be useful

HowardvanRooijen commented on content/03_CreatingObservableSequences.md:
A real-world scenario would be useful

idg10 replied:
Again, I can't tell which of the two sections spanned by this highlight you mean. If it's `AsyncSubject<T>` (which is what the first 3 lines are from) I've added one.

If it's about the subject factory, this is another case where I don't really know what this is for. It's a section I've retained from Lee's book. I've added some detail because it wasn't even clear to me what it did from his original text.

I'm slightly reluctant to remove it simply because for many years I've had a suspicion that there's some epiphany around subjects that has thus far eluded me. I know mechanically what they do and I know some uses for some of the concrete ones. But I never really understood why they are apparently imbued with particular significance in Reaqtor.

HowardvanRooijen reviewed (commented):
03 Reviewed.
I agree with Matthew's comment about full stops vs em-dash. Shorter sentences would be nice, even if the concepts do flow between sentences.

HowardvanRooijen reviewed (commented):
03 Reviewed.
I agree with Matthew's comment about full stops vs em-dash. Shorter sentences would be nice, even if the concepts do flow between sentences.

HowardvanRooijen commented on content/05_Filtering.md:
````diff
@@ -1,505 +1,427 @@
+Where completed
+```
+
+The `Where` operator is one of the many standard LINQ operators. This is one of many standard LINQ operators you'll find on all LINQ providers—LINQ to Objects, the `IEnumerable<T>` implementation, provides an equivalent method, for example. In most cases, Rx's operators behave just as they do in the `IEnumerable<T>` implementations, although there are some exceptions as we'll see later. We will d ...
````
Weird repetition.

HowardvanRooijen commented on content/05_Filtering.md:
```diff
@@ -1,505 +1,427 @@
+
+A side effect of this cascading `Subscribe` is that `Where` (and most other LINQ operators) is neither inherently _hot_ or _cold_: since it just subscribes to its source, then it will be hot if its source is hot, and cold if its source is cold.
+
+The `Where` operator passes on all elements for which its `predicate` callback returns `true`. To be more precise, when you subscribt to `Where`, it will create its own `IObserver<T>` which it passes as the argument to `source.Subscribe`, and this observer invokes the `predicate` for each call to `OnNext`. If that predicate returns `true`, then and only then will the observer created by `Where` c ...
```
Typo "subscribt"

HowardvanRooijen commented on content/05_Filtering.md:
```diff
@@ -1,505 +1,427 @@
+
+This logically has the same effect, but because we're using `await`, this won't block the calling thread while it waits for the observable source to produce a value. This might reduce the chances of deadlock.
+
+The fact that we're able to use `await` makes some sense of the fact that these methods end with `Async`, but you might be wondering what's going on here. We've seen that these methods all return `IObservable<T>`, not `Task<T>`, so how are we able to use `await`? There's a full explanation in the [Leaving Rx's World](13_LeavingIObservable.md) chapter, but the short answer is that Rx provides exte ...
```
It would be quite nice to have a bit more technical detail about how this works - is it using the awaiter approach where you can make anything awaitable (like how foreach works)?

idg10 replied:
As it says, there is a full explanation in a different chapter. Since this chapter is about filtering, it seemed to make more sense to put the description of how this works in the chapter that is already dedicated to integration between Rx and non-Rx stuff.

I've modified the link to point to the relevant section of the chapter, but I think that's enough—I don't think it would be an improvement to move that content out of that chapter and into this one.

> It would be quite nice to have a bit more technical detail about how this works - is it using the awaiter approach where you can make anything awaitable (like how foreach works)?

I'm not sure what you mean, because `await foreach` uses a different mechanism from all other `await` expressions. I _think_ (but I'm not sure) maybe you're referring to the pattern-based approach: `GetAwaiter` for `await` is a pattern just like `GetEnumerator` is a pattern for `foreach`, and all of the LINQ query expression stuff also expects similar well-known method names.

If that's what you meant, then yes. I don't want to put that in a chapter on filtering, because we're already one digression deep by talking about `await`, and this would takes us further away. I think it could go in the part where I discuss how `await` integration works, in Ch13's "Integration with `async` and `await`" section, but even there, my gut feeling is that drawing a comparison with `foreach` would be more distracting than illuminating.

HowardvanRooijen commented on content/05_Filtering.md:
````diff
@@ -1,505 +1,427 @@
+There is yet another standard LINQ operator for selecting one particular element from the source: `ElementAt`. You provide this with a number indicating the position in the sequence of the element you require. In data-at-rest LINQ providers, this is logically equivalent to accessing an array element by index. Rx implements this operator, but whereas most LINQ providers' `ElementAt<T>` implementat ...
+
+```cs
+IAisMessage fourth = await recieverHost.Message.ElementAt(4);
````
Typo "reciever" -> "receiver"

HowardvanRooijen commented on content/05_Filtering.md:
```diff
@@ -1,505 +1,427 @@
+- Handle the OnError gracefully
+- Use `.Skip(5).Take(1);` This will ignore the first 5 values and the only take the 6th value. 
+If the sequence has less than 6 elements we just get an empty sequence, but no errors.
+- Use ElementAtOrDefault
```
Should this be formatted as `ElementAtOrDefault`?

HowardvanRooijen commented on content/00_Foreword.md:
```diff
@@ -1,185 +1,105 @@
+
+The version that this book has been written against is `System.Reactive` version 6.0.
+
+So, fire up Visual Studio and let's get started.
```
Do you think we should add an instruction that if anyone finds any bugs / errata they should post an issue (or PR) at https://github.com/dotnet/reactive/tree/main/Rx.NET/Documentation/IntroToRx

HowardvanRooijen reviewed (commented):
05 reviewed.

idg10 committed:
Ch01 HvR feedback (except for backpressure)

idg10 committed:
Add short back pressure section

HowardvanRooijen commented on content/06_Transformation.md:
```diff
@@ -0,0 +1,351 @@
+
+## Cast
+
+C#'s type system is not omniscient. Sometimes we might know something about the type of the values emerging from an observable source that is not reflected in that source's type. This might be based on domain-specific knowledge. For example, with the AIS message broadcast by ships, we might know that if the message type is 3, it will contain navigation information. That means we could write this:
```
Does AIS use the parlance of ships or vessels? I say because there are often static platforms / sea planes etc

idg10 replied:
This was a deliberate bit of weasel wording. I was avoiding the word "vessel" on purpose.

The fact that AIS messages can also be broadcast by things that are not ships is true, but doesn't make what I've written untrue. This paragraph is under no obligation to provide a complete description of everything that an AIS message conceivably could be. It's OK just to talk about what happens when a ship broadcasts an AIS message.

Given that it's OK to say "ship" why do I prefer that to "vessel" here? The short answer is: if you're looking for a concrete example (and I believe you are in favour of concrete examples) more concrete is generally better (for pretty much the same reason a concrete example is better than no concrete example).

It is true that AIS uses the more general language of "vessels". And if you're in a maritime frame of mind, that's clear enough. But I felt that for anyone who had landed directly at this part of the docs by searching, and wasn't already familiar with the AIS scenarios I use throughout, it might not be obvious what a "vessel" is.

It's _too_ general. Coffee cups are a kind of vessel.

I felt that "AIS messages broadcast by ships" would be more self-evidently clear to a wide audience.

HowardvanRooijen commented on content/06_Transformation.md:
```diff
@@ -0,0 +1,351 @@
+
+The `Materialize` operator transforms a source of `IObservable<T>` into one of type `IObservable<Notification<T>>`. It will provide one `Notification<T>` for each item the source produces, and, if the sourced terminates, it will produce one final `Notification<T>` indicating whether it completed successfully or with an error.
+
+This can be useful because it produces objects that describe a whole sequence. If you wanted to record the output of an observable in a way that could later be replayed...well you'd probably use a `ReplaySubject<T>` because it is designed for precisely that job. But if you wanted to be able to inspect or modify the items in the sequence once they had been collectors, you might want to write your  ...
```
I'm not sure what is meant by "once they had been collectors"

HowardvanRooijen commented on content/00_Foreword.md:
```diff
@@ -1,185 +1,105 @@
+Duncan Mole, 
+Cathal Golden, 
+[Keith Woods](http://keith-woods.com), 
+[Ray Booysen](http://nondestructiveme.com) &amp; [Olivier DeHeurles](http://odeheurles.com/),
```
http://nondestructiveme.com no longer exists

HowardvanRooijen replied:
http://odeheurles.com/ no longer exists

HowardvanRooijen commented on content/00_Foreword.md:
```diff
@@ -1,185 +1,105 @@
+[John Marks](http://johnhmarks.wordpress.com/), 
+Duncan Mole, 
+Cathal Golden, 
+[Keith Woods](http://keith-woods.com), 
```
http://keith-woods.com/ on longer exists

HowardvanRooijen commented on content/00_Foreword.md:
```diff
@@ -1,185 +1,105 @@
+[Matt Davey](http://mdavey.wordpress.com), [Joe Albahari](http://www.albahari.com/) 
+and Gregory Andrien.
+Extra special thanks to the team at Microsoft that did the hard work and brought us Rx; 
+[Jeffrey Van Gogh](http://blogs.msdn.com/b/jeffva/), 
```
http://blogs.msdn.com/b/jeffva/ - 404

HowardvanRooijen commented on content/00_Foreword.md:
```diff
@@ -1,185 +1,105 @@
+and Gregory Andrien.
+Extra special thanks to the team at Microsoft that did the hard work and brought us Rx; 
+[Jeffrey Van Gogh](http://blogs.msdn.com/b/jeffva/), 
+[Wes Dyer](http://blogs.msdn.com/b/wesdyer/), 
```
http://blogs.msdn.com/b/wesdyer/ - 404

HowardvanRooijen commented on content/00_Foreword.md:
```diff
@@ -1,185 +1,105 @@
+[Jeffrey Van Gogh](http://blogs.msdn.com/b/jeffva/), 
+[Wes Dyer](http://blogs.msdn.com/b/wesdyer/), 
+[Erik Meijer](http://www.applied-duality.com/) &amp; 
+[Bart De Smet](http://blogs.bartdesmet.net/bart/). 
```
http://bartdesmet.net/ - 503

HowardvanRooijen reviewed (commented):
Just checking the links in the foreword - most of them are dead.

idg10 committed:
Ch02 - processed more feedback

Also updated a .NET 7 ref to in another chapter to .NET 8

idg10 committed:
Use fewer em dashes

idg10 committed:
Ch03 updates after feedback

idg10 committed:
Normalize use of generic type names in headings

idg10 committed:
Add missing code fence type

HowardvanRooijen commented on content/08_Partitioning.md:
```diff
@@ -0,0 +1,497 @@
+
+I've replaced `Select` with [`SelectMany`, also described in the Transformation chapter](06_Transformation.md#selectmany). As you may recall, `SelectMany` flattens nested observables back into a single flat sequence. You can see this reflected in the return type: now we've got just an `IObservable<IAisMessageType1to3>` instead of a sequence of sequences.
+
+Wait a second! Haven't I just undone the work that `GroupBy` did? I asked it to partition the events by vessel id, so why am I now recombining it back into a single, flat stream? Isn't that what I started with?
```
I love the real world example, but I wonder if this section would benefit from a marble diagram to help visualise the example?

HowardvanRooijen commented on content/08_Partitioning.md:
```diff
@@ -0,0 +1,497 @@
+
+The critical thing to understand here is that in the space of ten minutes, `receiverHost.Messages` produced _thousands_ of messages. (The rate varies by time of day, but it's typically over a thousand messages a minute. The code would have processed roughly ten thousand messages when I ran it to produce that output.) But as you can see, `shipStatusChanges` produced just 19 messages.
+
+This shows how Rx can tame high volume event sources in ways that are much more powerful than mere aggregation. We've not just reduced the data down to some statistical measure that can only provide an overview. Statistical measures such as averages or variance wouldn't be able to tell us anything about any particular ship. But here, every message tells us something about a particular ship. We've ...
```
The nuance that might be missing here is that you are applying real world business rules / domain rules / domain observations (or events) that are meaningful to the end user or downstream systems, rather than mathematical properties.

HowardvanRooijen commented on content/08_Partitioning.md:
```diff
@@ -0,0 +1,497 @@
+It can take thousands of messages and perform the necessary processing to find the handful that really matter to us.
+
+This is an example of the "fanning out, and then back in again" technique I described in ['The Significance of SelectMany' in the Transformation chapter](06_Transformation.md#the-significance-of-selectmany). This code uses `GroupBy` to fan out from a single observable to multiple observables. The key to this step is to create nested observables that provide the right level of detail for the proce ...
+
```
One example we have in the polyglot notebooks is to apply a geofence to an area of the Norwegian coast - this would be analagous to monitoring shipping vessels around a protected fishing area, where you could then track for suspicious activity.

HowardvanRooijen replied:
I'm not sure if this would be too much of a deep dive at this point, but it might be worth pointing out the additional memory / CPU overhead of using GroupBy - as you're capturing unbounded state (i.e. you don't know how many distinct vessels there will be), so a very simple query could be very resource intensive.

HowardvanRooijen commented on content/08_Partitioning.md:
```diff
@@ -0,0 +1,497 @@
+    perShipObservables.Select(shipMessages => shipMessages
+        .OfType<IAisMessageType1to3>()
+        .DistinctUntilChanged(m => m.NavigationStatus)
+        .Skip(1));
```
What's the purpose of `Skip(1)`?

HowardvanRooijen commented on content/08_Partitioning.md:
```diff
@@ -0,0 +1,497 @@
+
+Because `Buffer` returns an `IObservable<IList<T>>`, it can't produce a chunk until it has all of the elements that will go into that chunk. `IList<T>` supports random access—you can ask it how many elements it has, and you can retrieve any element by numeric index, and we expect these operations to complete immediately. (It would be technically possible to write an implementation of `IList<T>` r ...
+
+But because `Window` returns an observable that produces a nested observable to represent each chunk, it can emit that before necessarily having all of the elements. In fact, it emits a new window as soon as it knows it will need one. If you use `Window(4, 1)` for example, the observable it returns emits its first nested observable immediately. And then as soon as the source produces its first el ...
```
So is the difference really that Window fully embraces the Reactive paradigm, whereas Buffer is more IEnumerable?

idg10 replied:
That's half the story. I've chosen to add **two** sentences:

>You could say that `Window` more fully embraces the reactive paradigm than `Buffer`. Then again after some experience you might conclude that `Window` is harder to use than `Buffer` but is very rarely any more useful in practice.

At least, this has been my experience.

`Window` is in a sense the more fundamental operator, but `Buffer` is almost always the one you actually want to use. (I've yet to find a situation in which `Window` was the right choice. Abstractly I know it can do things that buffer can't but every time I've tried to come up with a practical example, I've concluded that actually `Window` wouldn't really help.

HowardvanRooijen commented on content/08_Partitioning.md:
````diff
@@ -0,0 +1,497 @@
+}
+```
+
+If we now want to extend this method to offer skip functionality, we need to have two different sequences: one for opening and one for closing. We open a window on subscription and again after the `skip` items have passed. We close those windows after '`count`' items have passed since the window opened.
````
Do you think there should be a call back to `Quiescent` in this section?

HowardvanRooijen reviewed (commented):
Chapter 08 Reviewed

idg10 committed:
Ch05 update after review comments

idg10 committed:
Ch05 process more review feedback

idg10 committed:
Ch06 updates after review feedback

idg10 committed:
Ch07 update after review

idg10 committed:
Ch08 update after review feedback

idg10 committed:
Ch11 fix Obserable typo

idg10 committed:
Replace spurious &lt; with <

idg10 committed:
Fix 'lamba' typo

idg10 committed:
Ch00 - update or remove broken links in acknowledgements

idg10 committed:
Ch00 add reference to https://reaqtive.net/ for Rx history

idg10 committed:
Ch00 fix more typos, add more links

idg10 committed:
Ch01 and Ch02 - more feedback based updates

idg10 committed:
Add thanks to open source contributors to Rx.

idg10 committed:
Add link to reactive x slack

idg10 committed:
Ch01 mention Reaqtor in paragraph on transactional queues

idg10 committed:
Ch01 use the word Telemetry to keep Howard happy

idg10 committed:
Ch16 fix some typos

idg10 committed:
App D - fix "howeever" typo

idg10 committed:
Change "favor" to UK spelling

idg10 committed:
Add thanks to .NET Foundation and Richard Lander

idg10 committed:
Remove spurious '.' in bullet list

idg10 committed:
Ch03 responding to mwa's feedback

idg10 committed:
Ch03 fix another typo

idg10 committed:
Change "ownership" to "stewards"

HowardvanRooijen commented on content/12_Timing.md:
```diff
@@ -0,0 +1,306 @@
+
+# Time-based sequences
+
+With event sources, timing is often important. In some cases, the only information of interest about some event might be the time at which it occurred. The core `IObservable<T>` and `IObserver<T>` interfaces don't mention timing at all in their method signatures, but they don't need to, because a source can decide when it calls an observer's `OnNext` method. A subscriber knows when an event occur ...
```
typo in ```[`Window`])``` - the `)` is preventing the link from being converted into  hyperlink

HowardvanRooijen commented on content/09_CombiningSequences.md:
```diff
@@ -0,0 +1,895 @@
+
+This marble diagram shows the items emerging from the two sources, `s1` and `s2`, and how `Concat` combines them into the result, `c`:
+
+![](./GraphicsIntro/Ch09-CombiningSequences-Marbles-Concat-Marbles.svg)
```
Are you going to add a description for this image?

HowardvanRooijen commented on content/09_CombiningSequences.md:
```diff
@@ -0,0 +1,895 @@
+
+Although Rx's `Concat` has the same logical behaviour as the [LINQ to Objects `Concat`](https://learn.microsoft.com/en-us/dotnet/api/system.linq.enumerable.concat), there are some Rx-specific details to be aware of. In particular, timing is often more significant in Rx than with other LINQ implementations. For example, in Rx we distinguish between [_hot_ and _cold_ source](02_KeyTypes.md#hot-and- ...
+
+![](./GraphicsIntro/Ch09-CombiningSequences-Marbles-Concat-Hot-Marbles.svg)
```
Are you going to add a description for this image?

HowardvanRooijen commented on content/09_CombiningSequences.md:
```diff
@@ -0,0 +1,895 @@
+
+This last example reveals that marble diagrams gloss over a detail: they show when a source starts, when it produces values, and when it finishes, but they ignore the fact that to be able to produce items at all, an observable source needs a subscriber. If nothing subscribes to an `IObservable<T>`, then it doesn't really produce anything. `Concat` doesn't subscribe to its second input until the f ...
+
+![](./GraphicsIntro/Ch09-CombiningSequences-Marbles-Concat-Hot-Marbles-SubOnly.svg)
```
Are you going to add a description for this image?

HowardvanRooijen commented on content/09_CombiningSequences.md:
```diff
@@ -0,0 +1,895 @@
+
+Most of the time we can get away with ignoring this subtlety, quietly using whichever convention suits us. To paraphrase [Humpty Dumpty: when I use a marble diagram, it means just what I choose it to mean—neither more nor less](https://www.goodreads.com/quotes/12608-when-i-use-a-word-humpty-dumpty-said-in-rather). But when you're combining hot and cold sources together, there might not be one obv ...
+
+![](./GraphicsIntro/Ch09-CombiningSequences-Marbles-Concat-Hot-Marbles-SourceAndSub.svg)
```
Are you going to add a description for this image?

HowardvanRooijen commented on content/09_CombiningSequences.md:
```diff
@@ -0,0 +1,895 @@
+
+We're using a distinct 'lane' in the marble diagram to represent the events seen by a particular subscription to a source. With this technique, we can also show what would happen if you pass the same cold source into `Concat` twice:
+
+![](./GraphicsIntro/Ch09-CombiningSequences-Marbles-Concat-Marbles-Cold-Twice.svg)
```
Are you going to add a description for this image?

HowardvanRooijen commented on content/09_CombiningSequences.md:
```diff
@@ -0,0 +1,895 @@
+
+Below is a marble diagram of the `Concat` operator applied to the `GetSequences` method. 's1', 's2' and 's3' represent sequence 1, 2 and 3. Respectively, 'rs' represents the result sequence.
+
+![](./GraphicsIntro/Ch09-CombiningSequences-Marbles-Concat-Marbles-Three.svg)
```
Are you going to add a description for this image?

HowardvanRooijen commented on content/09_CombiningSequences.md:
```diff
@@ -0,0 +1,895 @@
+
+To illustrate `Amb`'s behaviour, here's a marble diagram showing three sequences, `s1`, `s2`, and `s3`, each able to produce a sequence values. The line labelled `r` shows the result of passing all three sequences into `Amb`. As you can see, `r` provides exactly the same notifications as `s1`, because in this example, `s1` was the first sequence to produce a value.
+
+![](GraphicsIntro/Ch09-CombiningSequences-Marbles-Amb-Marbles.svg)
```
Are you going to add a description for this image?

HowardvanRooijen commented on content/09_CombiningSequences.md:
```diff
@@ -0,0 +1,895 @@
+
+If we changed the order so that `s2.OnNext(99)` came before the call to `s1.OnNext(1);` then s2 would produce values first and the marble diagram would look like this.
+
+![](GraphicsIntro/Ch09-CombiningSequences-Marbles-Amb-Marbles2.svg)
```
Are you going to add a description for this image?

HowardvanRooijen commented on content/09_CombiningSequences.md:
```diff
@@ -0,0 +1,895 @@
+
+Here is the marble diagram illustrating how this code behaves:
+
+![](./GraphicsIntro/Ch09-CombiningSequences-Marbles-Amb-Marbles3.svg)
```
Are you going to add a description for this image?

HowardvanRooijen commented on content/09_CombiningSequences.md:
```diff
@@ -0,0 +1,895 @@
+
+Since `Merge` returns a single observable sequence that includes all of the values from all of its input sequences, there's a sense in which it is similar to `Concat`. But whereas `Concat` waits until each input sequence completes before moving onto the next, `Merge` supports concurrently active sequences. As soon as you subscribe to the observable returned by `Merge`, it immediately subscribes t ...
+
+![](GraphicsIntro/Ch09-CombiningSequences-Marbles-Merge-Marbles.svg)
```
Are you going to add a description for this image?

HowardvanRooijen commented on content/09_CombiningSequences.md:
```diff
@@ -0,0 +1,895 @@
+
+As we can see from the marble diagram, s1 and s2 are yielded and subscribed to immediately. s3 is not yielded for one second and then is subscribed to. Once all input sequences have completed, the result sequence completes.
+
+![](GraphicsIntro/Ch09-CombiningSequences-Marbles-Merge-Marbles-Multi.svg)
```
Are you going to add a description for this image?

HowardvanRooijen commented on content/09_CombiningSequences.md:
```diff
@@ -0,0 +1,895 @@
+
+If we were lucky and each search completed before the next element from `searchValues` was produced, the output would look sensible. However, it is much more likely, however that multiple searches will result in overlapped search results. This marble diagram shows what the `Merge` function could do in such a situation.
+
+![](GraphicsIntro/Ch09-CombiningSequences-Marbles-Switch-Marbles-Bad-Merge.svg)
```
Are you going to add a description for this image?

HowardvanRooijen commented on content/09_CombiningSequences.md:
```diff
@@ -0,0 +1,895 @@
+Note how the values from the search results are all mixed together. The fact that some search terms took longer to get a search result than others has also meant that they have come out in the wrong order. This is not what we want. If we use the `Switch` extension method we will get much better results. `Switch` will subscribe to the outer sequence and as each inner sequence is yielded it will su ...
+
+
+![](GraphicsIntro/Ch09-CombiningSequences-Marbles-Switch-Marbles.svg)
```
Are you going to add a description for this image?

HowardvanRooijen commented on content/09_CombiningSequences.md:
```diff
@@ -0,0 +1,895 @@
+
+The effect can be seen in this marble diagram below.:
+
+![](GraphicsIntro/Ch09-CombiningSequences-Marbles-Zip-Marbles.svg)
```
Are you going to add a description for this image?

HowardvanRooijen commented on content/09_CombiningSequences.md:
```diff
@@ -0,0 +1,895 @@
+
+The marble diagram below shows off usage of `CombineLatest` with one sequence that produces numbers (`s1`), and the other letters (`s2`). If the `resultSelector` function just joins the number and letter together as a pair, this would produce the result shown on the bottom line. I've colour coded each output to indicate which of the two sources caused it to emit that particular result, but as you ...
+
+![](GraphicsIntro/Ch09-CombiningSequences-Marbles-CombineLatest-Marbles.svg)
```
Are you going to add a description for this image?

HowardvanRooijen commented on content/09_CombiningSequences.md:
```diff
@@ -0,0 +1,895 @@
+
+Let us now imagine a scenario where the left sequence produces values twice as fast as the right sequence. Imagine that in addition we never close the left windows; we could do this by always returning `Observable.Never<Unit>()` from the `leftDurationSelector` function. And imagine that we make the right windows close as soon as they possibly can, which we can achieve by making `rightDurationSele ...
+
+![](GraphicsIntro/Ch09-CombiningSequences-Marbles-Join-Marbles1.svg)
```
Are you going to add a description for this image?

HowardvanRooijen commented on content/09_CombiningSequences.md:
```diff
@@ -0,0 +1,895 @@
+
+What if I wanted to ensure that items from `right` only ever intersected with a single value from `left`? In that case, I'd need to ensure that the left durations did not overlap. One way to do that would be to have my `leftDurationSelector` always return the same sequence that I passed as the `left` sequence. This will result in `Join` making multiple subscriptions to the same source, and for so ...
+
+![](GraphicsIntro/Ch09-CombiningSequences-Marbles-Join-Marbles2.svg)
```
Are you going to add a description for this image?

HowardvanRooijen commented on content/09_CombiningSequences.md:
```diff
@@ -0,0 +1,895 @@
+
+This diagram shows those same inputs again, and also shows the observables `GroupJoin` would pass to the `resultSelector` for each of the items produced by `left`:
+
+![](GraphicsIntro/Ch09-CombiningSequences-Marbles-GroupJoin-Marbles.svg)
```
Are you going to add a description for this image?

HowardvanRooijen commented on content/09_CombiningSequences.md:
```diff
@@ -0,0 +1,895 @@
+
+![](./GraphicsIntro/Ch09-CombiningSequences-Marbles-Concat-Hot-Marbles-SubOnly.svg)
+
+This makes it easier to see why `Concat` produces the output it does. But since `hot` is a hot source here, this diagram fails to convey the fact that `hot` is producing items entirely on its own schedule. In a scenario where `hot` had multiple subscribers, then the earlier diagram would arguably be better because it correctly reflects every event available from `hot` (regardless of however many  ...
```
Missing "to" from 
> _starting from whenever they happened subscribe_.

HowardvanRooijen commented on content/09_CombiningSequences.md:
````diff
@@ -0,0 +1,895 @@
+}
+```
+
+We can call this `GetSequences` method and pass the results to `Concat`, and then use our `D ump` extension method to watch what happens:
````
Typo 
> D ump

HowardvanRooijen commented on content/09_CombiningSequences.md:
````diff
@@ -0,0 +1,895 @@
+}
+```
+
+`StartWith` might remind you of [`BehaviorSubject<T>`](03_CreatingObservableSequences.md#behaviorsubject), because that also ensures that consumers receive a value as soon as they subscribe. It's not quite the same: `BehaviorSubject<T>` caches the last value its own source emits. You might thank that would make it a better way to implement this vessel navigation example. However, since this examp ...
````
Typo -> think
> You might thank

HowardvanRooijen commented on content/09_CombiningSequences.md:
```diff
@@ -0,0 +1,895 @@
+
+Why is that useful?
+
+A common use case for `Amb` is when you want to produce some sort of result as quickly as possible, and you have multiple options for obtaining that result, but you don't know in advance which will be fastest. Perhaps there are multiple servers that could all potentially give you the answer you want, and it's impossible to predict which will have the lowest response time. You could send requests  ...
```
Is this analagous to `Task.WhenAny`?

HowardvanRooijen commented on content/09_CombiningSequences.md:
````diff
@@ -0,0 +1,895 @@
+IObservable<IAisMessage> allMessages = station1.Merge(station2);
+```
+
+If you want to combine more than two sources, you have a few options::
````
Typo ::

HowardvanRooijen commented on content/09_CombiningSequences.md:
```diff
@@ -0,0 +1,895 @@
+
+Each time a left duration window intersects with a right duration window, we get an output. The right duration windows are all effectively of zero length, but this doesn't stop them from intersecting with the left duration windows, because those all never end. So the first item from right has a (zero-length) window that falls inside two of the windows for the `left` items, and so `Join` produces  ...
+
+If I also immediately closed the left window by returning `Observable.Empty<Unit>`, or perhaps `Observable.Return(0)`, windows would never overlap, so no pairs would ever get produced. (In theory if both left and right produce items at _exactly_ the same time, then perhaps we might get a pair, but since the timing of events is never absolutely precise, it would be a bad idea to design a system th ...
```
Missing "the"?
> windows would never overlap

HowardvanRooijen commented on content/09_CombiningSequences.md:
```diff
@@ -0,0 +1,895 @@
+
+If I also immediately closed the left window by returning `Observable.Empty<Unit>`, or perhaps `Observable.Return(0)`, windows would never overlap, so no pairs would ever get produced. (In theory if both left and right produce items at _exactly_ the same time, then perhaps we might get a pair, but since the timing of events is never absolutely precise, it would be a bad idea to design a system th ...
+
+What if I wanted to ensure that items from `right` only ever intersected with a single value from `left`? In that case, I'd need to ensure that the left durations did not overlap. One way to do that would be to have my `leftDurationSelector` always return the same sequence that I passed as the `left` sequence. This will result in `Join` making multiple subscriptions to the same source, and for so ...
```
This doesn't made sense to me:
> and for some kinds of courses that might introduce unwanted side effects
Is "courses" a typo or am I missing something?

HowardvanRooijen reviewed (commented):
Chapter 09 reviewed.

HowardvanRooijen commented on content/11_SchedulingAndThreading.md:
This probably shouldn't use a `cs` info_string

HowardvanRooijen commented on content/11_SchedulingAndThreading.md:
Still not a fan of the em-dash.

HowardvanRooijen commented on content/11_SchedulingAndThreading.md:
Should there be a link to the "testing rx" chapter?

HowardvanRooijen replied:
Or to the `Test Schedulers` section on line 475 (which then references the testing rx chapter)

HowardvanRooijen commented on content/11_SchedulingAndThreading.md:
>will be no, so it will
The so feels out of place

idg10 replied:
Really? The sentence makes no sense if I remove it. Something needs to signify that the remainder of the sentence is a consequence of the answer being no.

I've put the no in quotes, which I think makes its proximity to "so" look less odd.

HowardvanRooijen commented on content/11_SchedulingAndThreading.md:
Can you elaborate on "some scenarios"?

idg10 replied:
No, but I have elaborated on why with the current state of the .NET docs it is no longer possible to say with any certain what the benefits might be. I've recommended testing against your actual workload.

HowardvanRooijen commented on content/11_SchedulingAndThreading.md:
> this sort of scheduler

these type of schedulers?

idg10 replied:
The start of the paragraph identifies a single category:

> schedulers that virtualize time

All the schedulers listed belong to this one category. I wanted to emphasize that these are all variations on a single theme, so I believe the use of the singular is best here. (Plural would be acceptable, but would lose the nuance I wanted.)

HowardvanRooijen commented on content/11_SchedulingAndThreading.md:
I'm wondering it would be possible to prefix the thread `Thread` or `T` because (maybe I'm tired), but I didn't actually spot the thread in the output on first scan!

HowardvanRooijen replied:
I'm wondering if we could use a slightly tweaked format, either something like:
```
[1] Main thread
[12] Created thread for EventLoopScheduler
[6] 21/07/2023 14:57:21: Main thread exiting
[6] 21/07/2023 14:57:22: Tick 0
[6] 21/07/2023 14:57:23: Tick 1
[6] 21/07/2023 14:57:24: Tick 2
```
or
```
Main thread [1]
Created thread for EventLoopScheduler [12]
21/07/2023 14:57:21: Main thread exiting [6]
21/07/2023 14:57:22: Tick 0 [6]
21/07/2023 14:57:23: Tick 1 [6]
21/07/2023 14:57:24: Tick 2 [6]
```

HowardvanRooijen commented on content/11_SchedulingAndThreading.md:
Wouldn't you use `nameof` rather than magic strings now?

HowardvanRooijen commented on content/11_SchedulingAndThreading.md:
The next few paragraphs are a big long and turn into a bit of a wall of text. I think they could do with being broken into a few smaller paragraphs

HowardvanRooijen commented on content/11_SchedulingAndThreading.md:
So is the `this` manifested as the discard?

idg10 replied:
I don't understand the question. I'm not sure how to change the text to prevent a misunderstanding because I can't work out what you think is happening here.

The discarded parameter here is where the `state` gets passed back into our callback. In this case we've passed the `Schedule` method a state of `null`. So the thing being discarded here is `null`.

I've realised that one of the examples this section showed was a bit pointless, so I've removed it, and modified the discussion to fit, so I'm hoping that with that, it will be easier to follow what's going on.

HowardvanRooijen replied:
> But perhaps we could avoid that weirdness of passing our own `this` argument:

I don't see "our own `this` argument" - where is it in the code???

idg10 replied:
This is a reference to the preceding example, which looks like this:

```cs
var next = scheduler.Schedule(this, static (innerScheduler, @this) => @this.LoopRec(innerScheduler));
```

I've tweaked the text again to make it clear that I'm referring to the earlier example.

HowardvanRooijen commented on content/11_SchedulingAndThreading.md:
Is it worth mentioning that this is non-durable?

HowardvanRooijen commented on content/11_SchedulingAndThreading.md:
Just a personal preference, but this feels like a use case for going from `var` style to new style. i.e.
`CancellationTokenSource tokenSource = new();`

HowardvanRooijen commented on content/11_SchedulingAndThreading.md:
Could the `var` be changed to `IDisposable` for clarity?

HowardvanRooijen reviewed (commented):
Chapter 11 reviewed.

HowardvanRooijen commented on content/12_Timing.md:
```diff
@@ -0,0 +1,306 @@
+
+As you can see from the output, the timings are not exactly one second but are pretty close. Some of this will be measurement noise in the `TimeInterval` operator, but most of this variability is likely to arise from the `Observable.Interval` class. There will always be a limit to the precision with which a scheduler can honour the timing request of it. Some scheduler introduce more variation tha ...
+
+In fact, the inherent variations in timing can make `Timestamp` particularly useful. The problem with simply looking at `DateTimeOffset.Now` is that it takes a non-zero amount of time to process an event, so you'll likely see a slightly different time each time you try to read the current time during the processing of one event. By attaching a timestamp once, we capture the time at which the even ...
```
Similar to how an AIS Base Station appends the timestamp!

HowardvanRooijen commented on content/13_LeavingIObservable.md:
````diff
@@ -0,0 +1,511 @@
+
+```cs
+IObservable<long> source = Observable.Interval(TimeSpan.FromSeconds(1))
+                       .Take(5);
````
Could `Take` go on the same line, as its total length would be less than the next line?

HowardvanRooijen commented on content/13_LeavingIObservable.md:
````diff
@@ -0,0 +1,511 @@
+could not sabotage
+```
+
+While I have used words like 'evil' and 'sabotage' in these examples, it is more often than not an oversight rather than malicious intent that causes problems. The failing falls first on the programmer who designed the leaky class. Designing interfaces is hard, but we should do our best to help consumers of our code fall into [the pit of success](https://learn.microsoft.com/en-gb/archive/blogs/br ...
````
Hanlon's Razor!

HowardvanRooijen reviewed (commented):
Chapter 13 Reviewed

HowardvanRooijen commented on content/15_PublishingOperators.md:
```diff
@@ -0,0 +1,409 @@
+
+# Publishing operators
+
+Hot sources need to be able to deliver events to multiple subscribers. While we can implement the subscriber tracking ourselves, it can be easier to write a degenerate source that works only for a single subscriber, and then to use one of Rx's _multicast_ operators to publish it as a multi-subscriber hot source. The example in ["Representing Filesystem Events in Rx"](03_CreatingObservableSequence ...
```
> degenerate source

Seems like an odd turn of phrase.

idg10 replied:
Really? Are you unfamiliar with its use as a technical term? (I thought it was a bog standard technical term, but perhaps it's a bit computer sciency and I hadn't realised.) Or is that you are familiar with that usage, but don't think it's the right thing here?

HowardvanRooijen replied:
I don't think I've seen that phrase before... hence why I flagged it. If I google it... the 6th result is... this page!

idg10 replied:
It's not really "degenerate source" as a phrase, so much as the word "degenerate" applied to, well, anything really.

In a technical context it tends to mean something that technically just about qualifies as an example of something more complex, but only barely. An empty set is a degenerate set. Adding together no values at all to produce a grand total of zero is a degenerate example of calculating a sum total. The number zero can, strictly speaking, be considered to be a 14-degree polynomial, but it's one where all 14 coefficients are zero, so it's a degenerate 14-degree polynomial. A completely ordinary number qualifies, technically, as a kind of tensor, but one where all the things that make tensors interesting have effectively been turned down to zero, so although it does meet all the defining qualifications, it's a degenerate tensor.

It comes up in maths all the time because it can often be useful to treat something very simple as an example (albeit a degenerate example) of something more complex.

idg10 replied:
...but since the term evidently isn't as widely understood as I thought, I've replaced it. (And the amount of text it took to replace it is the reason I used the word in the first place.)

HowardvanRooijen commented on content/15_PublishingOperators.md:
````diff
@@ -0,0 +1,409 @@
+Because `Subject<T>` forwards all incoming `OnNext` calls to each of its subscribers immediately, and because it doesn't store any previously made calls, the result is a hot source. If you attach some subscribers before calling `Connect`, and then you attached more subscribers after calling `Connect`, those later subscribers will only receive events that occurred after they subscribed. This examp ...
+
+```cs
+IConnectableObservable<long> pticks = Observable
````
Why `pticks`?

idg10 replied:
In a section called **P**ublish, and where we've taken some ticks and **p**ublished them, I had hoped readers would infer what the **p** was short for.

I've renamed it `publishedTicks` just to be absolutely clear.

HowardvanRooijen commented on content/15_PublishingOperators.md:
```diff
@@ -0,0 +1,409 @@
+
+The `AutoConnect` operator behaves in much the same way as `RefCount`, in that it calls `Connect` on its underlying `IConnectableObservable<T>` when the first subscriber subscribers. The difference is that it doesn't attempt to detect when the number of active subscribers has dropped to zero: once it connects, it remains connected indefinitely, even if it has no subscribers.
+
+Although `AutoConnect` can be convenient, you need to be a little careful, because it can cause leaks: it will never disconnect automatically. It is still possible to tear down the connection it creates: `AutoConnect` accepts an optional argument of type `Action<IDiposable>`. It invokes this when it first connects to the source, passing you the `IDisposable` returned by the source's `Connect` met ...
```
Typo `IDiposable`
> `Action<IDiposable>`

HowardvanRooijen commented on content/15_PublishingOperators.md:
````diff
@@ -0,0 +1,409 @@
+Sub3: 4 (10/08/2023 16:40:41)
+Sub4: 4 (10/08/2023 16:40:41)
+```
+
````
This could do with an example for completeness.

HowardvanRooijen reviewed (commented):
Chapter 15 reviewed

HowardvanRooijen commented on content/16_TestingRx.md:
```diff
@@ -1,846 +1,722 @@
+
+Note that the output is almost the same. Scheduling of the creation and subscription do not affect the Hot Observable, therefore the notifications happen 1 tick earlier than their Cold counterparts.
+
+We can see the major difference a Hot Observable bears by changing the virtual create time and virtual subscribe time to be different values. With a Cold Observable, the virtual create time has no real impact, as subscription is what initiates any action. This means we can not miss any early message on a Cold Observable. For Hot Observables, we can miss messages if we subscribe too late. Here, we ...
```
> This means we can not miss

cannot?

HowardvanRooijen commented on content/16_TestingRx.md:
````diff
@@ -1,846 +1,722 @@
+OnCompleted() @ 40000000
+```
+
+### CreateObserver 
````
Provide a code sample?

idg10 replied:
I think that might be better done as part of a rethink, incorporating some practical examples from the repo itself. (This was a chapter where I didn't change very much in this pass.)

So I've set up https://github.com/dotnet/reactive/issues/2069

HowardvanRooijen reviewed (commented):
Chapter 16 reviewed

HowardvanRooijen commented on content/02_KeyTypes.md:
```diff
@@ -1,488 +1,629 @@
+And some examples of some sources that might make good cold observables:
+
+* the contents of a collection (such as is returned by the [`ToObservable` extension method for `IEnumerable<T>`](03_CreatingObservableSequences.md#from-ienumerablet))
+* a fixed range of values, such as [`Observable.Range`](03_CreatingObservableSequences.md#observablerange) produces
```
> 03_CreatingObservableSequences.md#observablerange

Should be 03_CreatingObservableSequences.md#observable.range

idg10 replied:
Should it? VS code thinks it's right as it is. If I add the full stop as you suggest, the link is now broken inside VS Code.

Is this another "opinion is divided on the matter" markdownism?

HowardvanRooijen replied:
It might be two conflicting approaches.

The Id generated by that heading is https://www.introtorx.com/chapters/creating-observable-sequences#observable.create

HowardvanRooijen replied:
and is defined by markdig as "auto-identifiers" https://github.com/xoofx/markdig/blob/master/src/Markdig.Tests/Specs/AutoIdentifierSpecs.md

HowardvanRooijen commented on content/02_KeyTypes.md:
```diff
@@ -1,488 +1,629 @@
+
+* the contents of a collection (such as is returned by the [`ToObservable` extension method for `IEnumerable<T>`](03_CreatingObservableSequences.md#from-ienumerablet))
+* a fixed range of values, such as [`Observable.Range`](03_CreatingObservableSequences.md#observablerange) produces
+* events generated based on an algorithm, such as [`Observable.Generate`](03_CreatingObservableSequences.md#observablegenerate) produces
```
> 03_CreatingObservableSequences.md#observablegenerate

Should be 03_CreatingObservableSequences.md#observable.generate

HowardvanRooijen commented on content/02_KeyTypes.md:
```diff
@@ -1,488 +1,629 @@
+* a fixed range of values, such as [`Observable.Range`](03_CreatingObservableSequences.md#observablerange) produces
+* events generated based on an algorithm, such as [`Observable.Generate`](03_CreatingObservableSequences.md#observablegenerate) produces
+* a factory for an asynchronous operation, such as [`FromAsync`](03_CreatingObservableSequences.md#from-task) returns
+* events produced by running conventional code such as a loop; you can create such sources with [`Observable.Create`](03_CreatingObservableSequences.md#observablecreate)
```
> 03_CreatingObservableSequences.md#observablecreate

Should be 03_CreatingObservableSequences.md#observable.create

HowardvanRooijen commented on content/02_KeyTypes.md:
````diff
@@ -1,488 +1,629 @@
+But how does `outstanding` itself work? The basic approach here is that every time `source` produces a value, we actually create a brand new `IObservable<int>`, which produces exactly two values. It immediately produces the value 1, and then after the specified timespan (2 seconds in these examples) it produces the value -1. That's what's going in in this clause of the query expression:
+
+```cs
+from delta in Observable.Return(1, scheduler).Concat(Observable.Return(-1, scheduler).Delay(minimumInactivityPeriod, scheduler))
````
Can this be formatted to go onto multiple lines (for ebook formatting)

idg10 committed:
Add sequence diagram to Ch03

idg10 committed:
Ch03 Add short section on long-term state in operators

idg10 committed:
Appendix A - fix awkward wording

idg10 committed:
Use markdown in Appendix B.

Inexplicably, this chapter was written mostly in HTML in a `.md` file, which confused some parts of our tooling. I've recast it all as markdown.

I've also written a short note to say you can use most of these disposable helpers in non-Rx code too.

idg10 committed:
Replace a tab with spaces

idg10 committed:
Appendix D: fix some typos

idg10 committed:
Ch12 - fix markdown snafu

idg10 committed:
Ch09 add some missing diagram descriptions.

idg10 committed:
Ch12 remove backticks in links because the toolchain is apparently confused

idg10 committed:
Ch12 trying Felix's suggestion for links to see if it renders OK on GitHub

idg10 committed:
Ch12 put links back how they were since neither GitHub nor VS Code liked Felix's suggestion

idg10 committed:
AppB: Remove spurious use of HTML escaping

idg10 committed:
Ch09 add remaining diagram descriptions and fix more typos

idg10 committed:
Ch11 updates after review feedback

idg10 committed:
Ch13 updates after feedback

idg10 committed:
Ch15 updates after feedback

idg10 committed:
Ch16 fix typo

idg10 committed:
Ch02 Split example across multiple lines to help ebook rendering

idg10 committed:
Try to clarify use of `@this` again

idg10 committed:
Ch15 remove word degenerate because apparently it's not as widely understood as I thought
